ENH: Use AVX for float32 implementation of np.sin & np.cos

r-devulap opened the pull request:
This commit implements vectorized single precision sine and cosine using
AVX2 and AVX512. Both sine and cosine are computed using a
polynomial approximation which are accurate for values between
[-PI/4,PI/4]. The original input is reduced to this range using a 3-step
Cody-Waite's range reduction method. This method is only accurate for
values between [-71476.0625f, 71476.0625f] for cosine and [-117435.992f,
117435.992f] for sine. The algorithm identifies elements outside this
range and calls glibc in a scalar loop to compute their output.

The algorithm is a vectorized version of the methods presented
here: https://stackoverflow.com/questions/30463616/payne-hanek-algorithm-implementation-in-c/30465751#30465751

Accuracy: maximum ULP error = 1.49

Max ULP error comparison to GLIBC's implementation: 

| Function | GLIBC scalar | GLIBC vector (libmvec) | this patch |               
|----------|--------------|------------------------|------------|               
| sine     | 1.19         | 2.47                   | 1.49       |               
| cosine   | 1.19         | 2.39                   | 1.49       |               

Performance: The speed-up this implementation provides is dependent on
the values of the input array. AVX512 version performs nearly 22x faster when all the input
values are within the range specified above. Details of the performance
benefits are provided in the commit message. Its worst performance is when all the array
elements are outside the range leading to about 1% reduction in
performance.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

Commit:
ENH: Use AVX for float32 implementation of np.sin & np.cos

This commit implements vectorized single precision sine and cosine using
AVX2 and AVX512. Both sine and cosine are computed using a
polynomial approximation which are accurate for values between
[-PI/4,PI/4]. The original input is reduced to this range using a 3-step
Cody-Waite's range reduction method. This method is only accurate for
values between [-71476.0625f, 71476.0625f] for cosine and [-117435.992f,
117435.992f] for sine. The algorithm identifies elements outside this
range and calls glibc in a scalar loop to compute their output.

The algorithm is a vectorized version of the methods presented
here: https://stackoverflow.com/questions/30463616/payne-hanek-algorithm-implementation-in-c/30465751#30465751

Accuracy: maximum ULP error = 1.49

Performance: The speed-up this implementation provides is dependent on
the values of the input array. It performs best when all the input
values are within the range specified above. Details of the performance
boost are provided below. Its worst performance is when all the array
elements are outside the range leading to about 1-2% reduction in
performance.

Three different benchmarking data are provided, each of which was benchmarked
using timeit package in python. Each function is executed 1000 times and
this is repeated 100 times. The standard deviation for all the runs was
less than 2% of their mean value and hence not included in the data.

(1) Micro-bencharking:
Array size = 10000, Command = "%timeit np.cos([myarr])"
|---------------+------------+--------+---------+----------+----------|
| Function Name | NumPy 1.16 | AVX2   | AVX512  | AVX2     | AVX512   |
|               |            |        |         | speed up | speed up |
|---------------+------------+--------+---------+----------+----------|
| np.cos        | 1.5174     | 0.1553 | 0.06634 | 9.77     | 22.87    |
| np.sin        | 1.4738     | 0.1517 | 0.06406 | 9.71     | 23.00    |
|---------------+------------+--------+---------+----------+----------|

(2) Package ai.cs provides an API to transform spherical coordinates to
cartesean system:
Array size = 10000, Command = "%timeit ai.cs.sp2cart(r,theta,phi)"
|---------------+------------+--------+--------+----------+----------|
| Function Name | NumPy 1.16 | AVX2   | AVX512 | AVX2     | AVX512   |
|               |            |        |        | speed up | speed up |
|---------------+------------+--------+--------+----------+----------|
| ai.cs.sp2cart | 0.6371     | 0.1066 | 0.0605 | 5.97     | 10.53    |
|---------------+------------+--------+--------+----------+----------|

(3) Package photutils provides an API to find the best fit of first and
second harmonic functions to a set of (angle, intensity) pairs:
Array size = 1000, Command = "%timeit fit_first_and_second_harmonics(E, data)"
|--------------------------------+------------+--------+--------+----------+----------|
| Function Name                  | NumPy 1.16 | AVX2   | AVX512 | AVX2     | AVX512   |
|                                |            |        |        | speed up | speed up |
|--------------------------------+------------+--------+--------+----------+----------|
| fit_first_and_second_harmonics | 1.598      | 0.8709 | 0.7761 | 1.83     | 2.05     |
|--------------------------------+------------+--------+--------+----------+----------|

r-devulap commented:
I realize the first 6 commits are exactly same as in this PR: https://github.com/numpy/numpy/pull/13134. The only commit that is new is the last one which implements sine and cosine (but it relies on the previous commits).

eric-wieser commented:
For reviewers - if the other PR is merged, you can switch the base of this PR between <anything> and back to `master`, and it will recompute the diff and throw out all the merged commits.

mattip commented:
#13134 is merged. Could you fix the conflicts/rebase/switch the base?

r-devulap commented:
Thanks! Its updated now and has just the one commit.

mattip commented:
How do current benchmarks compare before/after this PR?

r-devulap commented:
As mentioned in the commit message, the performance gains are leveraged only if the input arguments to sine/cosine are between [-71476.0625f, 71476.0625f] for cosine and [-117435.992f,
117435.992f] for sine.  I measured 3 different benchmarks.

(1) For micro-benchmarks: 
Array size = 10000, Command = "%timeit np.cos([myarr])"

| Function Name | NumPy 1.16 | AVX2   | AVX512  | AVX2 speed up   | AVX512  speed up |
| ------------- | ----------- | -------- | --------- | ---------- | ---------- |
| np.cos        | 1.5174s     | 0.1553s | 0.06634s | 9.77x     | 22.87x    |
| np.sin        | 1.4738s     | 0.1517s | 0.06406s | 9.71x     | 23.00x    |

(2) Package ai.cs provides an API to transform spherical coordinates to
cartesean system:
Array size = 10000, Command = "%timeit ai.cs.sp2cart(r,theta,phi)"

| Function Name | NumPy 1.16 | AVX2   | AVX512  | AVX2 speed up   | AVX512  speed up |
| ------------- | ----------- | -------- | --------- | ---------- | ---------- |
| ai.cs.sp2cart | 0.6371s     | 0.1066s | 0.0605s | 5.97x     | 10.53x    |

(3) Package photutils provides an API to find the best fit of first and
second harmonic functions to a set of (angle, intensity) pairs:
Array size = 1000, Command = "%timeit fit_first_and_second_harmonics(E, data)"

| Function Name | NumPy 1.16 | AVX2   | AVX512  | AVX2 speed up   | AVX512  speed up |
| ------------- | ----------- | -------- | --------- | ---------- | ---------- |
| fit_first_and_second_harmonics | 1.598s      | 0.8709s | 0.7761s | 1.83x     | 2.05x   |

mattip commented:
Sorry for not being clear. I meant that I would expect that a performance PR would come with changes in benchmarks in `numpy.benchmarks`, run by 
```
python runtests.py --bench
```

r-devulap commented:
As far as I could tell, those benchmarks were running the float64 version of sine and cosine. Is there a way I could run the float32 version of it? Or do they already do that?

r-devulap commented:
I will take a look at this today and let you know if I am able to run the benchmarks for float32.

r-devulap commented:
I ran the ufunc benchmarks `python  runtests.py --bench ufunc` and saw this: 

| Function | Before this patch | With this patch |
| -- | -- | -- |
| sin | 7.51±0.01ms | 7.15±0ms |
| cos | 7.73±0.02ms | 7.36±0ms | 

But the thing is, the benchmarks for ufunc seems to reporting just one metric (sum of the total time, I assume) for a whole bunch of dtypes (int16, int32, float16, float32, float128, complex64, complex256), whereas my patch only improves performance for float32. So I am not sure if these numbers are representative of performance of this patch.

I added my own benchmarks which measures computing sine and cosine for just float32: 

```
class Trignometry(Benchmark):                                                   
    def setup(self):                                                            
        self.f = np.ones(10000, dtype=np.float32)                           
                                                                                
    def time_sine(self):                                                        
        np.sin(self.f)                                                          
                                                                                
    def time_cosine(self):                                                      
        np.cos(self.f)  
```

and this is what I measured: 

| Function | Before this patch | With this patch |
| -- | -- | -- |
| sin | 117±0.04us | 12.2±1us |
| cos | 115±0.02us | 12.9±1us |

juliantaylor commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -1617,6 +1617,36 @@ FLOAT_@func@_@isa@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY
+#else
+    /*
+     * This is the path it would take if ISA was runtime detected, but not
+     * compiled for. It fixes the error on clang6.0 which fails to compile
```
you could add the failure to compile the ISA into the runtime check (so it always returns false)

mattip replied:
@r-devulap any thoughts?

r-devulap replied:
This isn't a problem anymore actually. The compile time and runtime checks use same preprocessor macro.

juliantaylor commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -1270,6 +1352,137 @@ static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ @vtype@
+
+#if defined HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS
+static NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ void
+@ISA@_sincos_FLOAT(npy_float * op, npy_float * ip, const npy_int array_size,
```
`npy_int` is a 32 bit integer so it will fail on very large arrays
please use npy_intp (that also applies to the already merged exp/log code)

r-devulap replied:
thanks for catching that. I updated it.

juliantaylor commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -1270,6 +1352,137 @@ static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ @vtype@
+    const npy_int num_lanes = @BYTES@/sizeof(npy_float);
+    npy_int compute_cos = 1;
+    npy_float large_number = 71476.0625f;
+    if (*operation == 's') {
```
an enum would be better to select the operation

mattip replied:
@r-devulap ping

r-devulap replied:
oh, sorry this almost slipped my mind. I will add an enum and fix this.

r-devulap replied:
Is there a reason why enum is better than char comparison? (just for my understanding)

eric-wieser replied:
Because the enum documents intent better, in this case by making it clear what the complete set of allowing values is

r-devulap replied:
thanks, added a commit to use enum instead.

mattip commented:
This makes `_multiarray_umath.so` on Ubuntu 18.04 with gcc-8 grow from 17271072 to 17299296, which is 27.5 kB or 0.16%.

Commit:
BUG: fixing NAN handling and adding tests for sin/cos

r-devulap commented:
Made some edits to handle NAN and added some tests as well.

r-devulap commented:
Added a commit which fixes the bug reported in https://github.com/numpy/numpy/issues/13493 for sin and cos.

Commit:
BUG: sin and cos cast float16 to float32

Commit:
TEST: adding tests to validate AVX sin/cos implementation

Commit:
BUG: use strides and process strided arrays using AVX

Commit:
BUG: AVX2 impl of sin/cos requires an FMA

Without an FMA, the output of AVX2 and AVX512 version differ. This
changes ensures the output across implementations remains exactly the
same.

Commit:
TEST: Rounding max tolerable ulp error to an int

The assert_array_max_ulp returns only an int since it compares ULP
difference between two float32 numbers.

r-devulap commented:
Updated with following fixes: 
- using AVX for strided arrays (fixes the same issue reported for exp/log)
- sin/cos will require an fma instead of just avx2. Otherwise the results are in-consistent across avx2 and avx512 implementations. 
- Rounded up the max tolerable ulp error to nearest int (the `assert_array_max_ulp `returns an int)

charris commented:
Unused variable warnings.

r-devulap commented:
yeah, thanks :) It was in cpuid.c, just fixed it.

Commit:
BUG: eliminate unsed variables warning in cpuid

Commit:
MAINT: using an enum to switch between sin/cos

r-devulap commented:
seems like some network issues resulted in the one test failing and shouldn't be related to the patch itself.

Commit:
TEST: improving test coverage for sin/cos for input > 117435.992f

r-devulap commented:
just checking if there is any more feedback for this patch?

r-devulap commented:
@mattip @juliantaylor ping ..

r-devulap commented:
Hi @mattip, just wanted to know what you are thinking with respect to this patch. I have more performance patches I am working on but they rely on some of the commits in this patch.

mattip commented:
I am OK with this, it does not add much bloat and gives a considerable performance boost. @juliantaylor any thoughts?

r-devulap commented:
ping @mattip  ...

mattip commented:
I am a bit torn on this. I would like to merge it, but would like to have gh-13515 in place (tests for floating point accuracy) first.

r-devulap commented:
> I am a bit torn on this. I would like to merge it, but would like to have [gh-13515](https://github.com/numpy/numpy/issues/13515) in place (tests for floating point accuracy) first.

sounds good, I will get a patch going and submit a separate PR for this. As pointed out by @juliantaylor, glibc has nice list of floating point numbers to test each function (example: [here](https://sourceware.org/git/?p=glibc.git;a=blob;f=math/auto-libm-test-out-cos;h=072e04f157123e5439995a35a4037f67ded2af1f;hb=HEAD)). I can get started with this list first.

r-devulap commented:
@mattip PR #14048 has the test suite covering float32 exp, log, sin and cos. Let me know if that looks good.

r-devulap commented:
Re-based with master. The AVX functions for sin and cos pass the tests on my local skylake machine (though it is currently disabled in the CI).

Commit:
BUG: rename avx2_scalef_ps to fma_scalef_ps

mattip commented:
@r-devulap what would it take to enable the tests in CI? Do we need to find a skylake machine?

r-devulap commented:
If you have a cpu > Haswell then the AVX2/FMA algorithms should already be tested. SkylakeX will be needed for AVX512F. Additionally, I also need a way to enable the test only on these platforms and ignore on the rest.

r-devulap commented:
@mattip ping, how do you want to proceed with PR? Is there a way to enable the tests only for AVX machines?

mattip commented:
My inclination would be to put this in as-is, since you made sure tests pass on the relevant systems. I am not sure I understand the problem. I see all the checks pass sucessfully. Which tests are enabled for AVX-only?

r-devulap commented:
The test `test_validate_transcendentals` in the file `numpy/core/tests/test_umath_accuracy.py` is currently disabled completely in the CI. The reason is that it is different platforms/compilers seem to generate different outputs for the same functions and the accuracy varies significantly. I would like to know if there is way to enable this test in the CI only for platforms that NumPy utilizes the AVX version of sin/cos/exp/log and ignore the test on other platforms.

mattip commented:
I guess you could add a function to `numpy/core/src/umath/_umath_tests.c` which would export a C-level detection of AVX, then call it as part of the setup for those tests.

Would you like to do that as part of this PR or in a separate, follow-on one?

r-devulap commented:
that might take me some time to implement, perhaps a separate follow up PR is better?

mattip commented:
OK, let's put this in as-is, in the hope we will improve #14048 's testing over time

mattip commented:
thanks @r-devulap

r-devulap commented:
> OK, let's put this in as-is, in the hope we will improve #14048 's testing over time

sounds good, thanks!

tylerjereddy commented:
I've done the SkylakeX CI emulation testing upstream in OpenBLAS for AVX-related testing purposes: https://github.com/xianyi/OpenBLAS/pull/2134

Not sure that NumPy would really want to go that far though.
